README.md
# keymage.js

Keymage is a small (1.6kb after Closure Compiler and gzip) library for handling
key bindings in JavaScript. It supports nested application scopes, has a simple
DSL for defining keys and can handle key chords.

[![Build Status](https://travis-ci.org/piranha/keymage.svg?branch=master)](https://travis-ci.org/piranha/keymage) - or [check tests](https://rawgit.com/piranha/keymage/master/test/test.html) in browser

## Features

 - Simple language for defining bindings
 - Key sequences (a-la Emacs chords)
 - Nested scopes
 - Default modifier (`defmod` key which is `command` on OS X and `control`
   elsewhere)
 - Ability to prevent defaults for whole sequence


## Usage

Include `keymage.min.js` in your page:

```html
<script src="keymage.min.js"></script>
```

There are no dependencies. It is possible to use library as a simple JS module,
as an AMD module or as CommonJS module.

It worth to note that [Keymage is on cdnjs](http://cdnjs.com/libraries/keymage/)
which enables you to use it without downloading.

Plus, of course, it's on [NPM](https://www.npmjs.org/package/keymage).

## Defining shortcuts

Keymage exposes a single function, `keymage`:

```javascript
// bind on 'a'
keymage('a', function() { alert("You pressed 'a'"); });

// returning false prevents default browser reaction (you can always use
// e.preventDefault(), of course)
keymage('ctrl-e', function() { return false; });

// binding on 'defmod' binds on Command key on OS X and on Control key in other
// systems
keymage('defmod-j', function() { alert("I am fired"); });
```

Handler function receives two arguments: the original event and the context so
you can understand what and why was fired.

The context contains those properties:

 - `shortcut` is a string you've originally provided for binding
 - `scope` is a scope which is currently active
 - `definitionScope` is a scope where this shortcut was defined

```javascript
keymage('alt-c', function(e, ctx) {
    console.log(ctx.shortcut, ctx.scope, ctx.definitionScope);
});

// -> "alt-c", "", ""
```

## Sequences

Keymage supports key sequences:

```javascript
keymage('ctrl-k j', function() { alert("Nice!"); });
```

For this to fire you have to first press both `ctrl` and `k`, and then
`j`. This will fire an alert.

There is something to remember: browsers have their own shortcuts, for example
`ctrl-j` in most browsers means "open downloads". And while you can always call
`e.preventDefault()` in usual situation, if `ctrl-j` is part of a sequence, it's
not that easy - you'll get control over what's happening only when the whole
sequence is pressed.

So keymage provides you with a means to support this use case. I do not
encourage you to override browser hotkeys, but let's imagine you want to do
that. For this, you can pass an option object as last parameter, having
`preventDefault` property set to `true`:

```javascript
keymage('ctrl-t ctrl-j k',
        function() { alert("wow"); },
        {preventDefault: true});
```

This option will prevent default on *every* key press which is a valid part of a
bound sequence, including the one triggering your handler. Note that pressing
only `ctrl-j` (without `ctrl-t`) will still open downloads, keymage looks for
sequence of `ctrl-t ctrl-j`.

## Scopes

Keymage support nested scopes. This means that your application can have few
areas where you can gradually have more and more specific shortcuts. It works
like this:

```javascript
// You can skip scope argument if you want global work-always shortcut
keymage('ctrl-j q', function() { alert("Default scope"); });

// This will fire after "keymage.setScope('chat')"
keymage('chat', 'ctrl-j w', function() { alert("Chat scope"); });

// This will fire after "keymage.setScope('chat.input')"
keymage('chat.input', 'ctrl-j e', function() { alert("Chat.input scope"); });
```

You can control scopes with helpful `pushScope` and `popScope` methods. This way
your nested view (or whatever is enabling nested scope) doesn't need to know
about parent scope:

```javascript
keymage.pushScope('chat') // scope is 'chat'

keymage.pushScope('input') // scope is 'chat.input'

keymage.popScope() // scope is 'chat'

keymage.pushScope('deep')
keymage.pushScope('deeper') // scope is 'chat.deep.deeper'

// way to jump out of deep scoping
keymage.popScope('chat') // scope is ''
```

`pushScope` returns resulting scope and `popScope` returns topmost scope it
removed (so with parameters it's the one you've asked to remove).

Note that calling `popScope` with name of a scope which is repeated few times
will pop topmost one, i.e.:

```javascript
keymage.setScope('this.scope.is.deep.scope')
keymage.popScope('scope') // scope is 'this'
```


## Options

Last and optional argument to `keymage` function is an option object. Here is a
list of possible options:

 - `preventDefault`: when `true`, calls `event.preventDefault()` on every key
   press which looks like a part of defined sequence.

 - `context`: binding handler will be called with provided object as a context.


## Unbinding

And if you ever need to unbind a handler, use this:

```javascript
keymage.unbind('ctrl-j k', your_handler_function);
```

Also, `keymage(...)` returns a function, which unbinds this shortcut when called:

```javascript
var unbinder = keymage('ctrl-j k', your_handler_function);
unbinder();
```


CHANGELOG.md
# keymage changelog

## 1.1.3

- `keymage('key chain', ...)` now returns function to unbind this shortcut

## 1.1.2

- Shortcuts with `-` are now allowed

## 1.1.0

- It's now possible to unbind

## 1.0.2

- Export in CommonJS env

## 1.0.1

- `keymage.version` contains current version of keymage
- Popping topmost scope without proving it's name was broken

## 1.0

- First stable version, works for me, etc :)


test/test.js
var MODS = ['ctrl', 'shift', 'alt', 'meta'];
function fire(key) {
    var e = document.createEvent('Event');

    if (typeof key === 'string') {
        key = keymage.parse(key);
    }

    e.initEvent('keydown', true, true);
    e.keyCode = key.code;
    for (var i = 0; i < MODS.length; i++) {
        var mod = MODS[i];
        if (key[mod]) {
            e[mod + 'Key'] = true;
        }
    }
    document.dispatchEvent(e);
}

wru.test([
    {name: 'Single shortcut',
     test: function() {
         var count = 0;
         keymage('a', function() { count++; });

         fire({code: 65});
         wru.assert('pressed', count === 1);

         fire({code: 65});
         wru.assert('pressed', count === 2);

         keymage('ctrl-alt--', function() { count--; });
         fire({code: 189, ctrl: true, alt: true});
         wru.assert('pressed', count === 1);
     }},

    {name: 'Sequence',
     test: function() {
         var count = 0;
         keymage('ctrl-a a', function() { count++; });

         fire({code: 65, ctrl: true}); fire({code: 65});
         wru.assert('Sequence works', count === 1);

         // failing attempts
         fire({code: 65, ctrl: true}); fire({code: 66});
         fire({code: 65, ctrl: true}); fire({code: 65, ctrl: true}); fire({code: 65});
         wru.assert('C-a b & C-a C-a a - do not trigger sequence', count === 1);

         // still works
         fire({code: 65, ctrl: true}); fire({code: 65});
         wru.assert('But it still works', count === 2);
     }},

    {name: 'Scopes',
     test: function() {
         var count = 0;
         var handler = function() { count++; };
         keymage('ctrl-a a', handler);
         keymage('chat', 'ctrl-a b', handler);
         keymage('chat.input', 'ctrl-a c', handler);

         keymage.pushScope('chat');
         fire({code: 65, ctrl: true}); fire({code: 65});
         wru.assert('C-a a - triggers, chat scope', count === 1);
         fire({code: 65, ctrl: true}); fire({code: 66});
         wru.assert('C-a b - triggers, chat scope', count === 2);
         fire({code: 65, ctrl: true}); fire({code: 67});
         wru.assert('C-a c - does not trigger, chat scope', count === 2);

         keymage.pushScope('input');
         fire({code: 65, ctrl: true}); fire({code: 67});
         wru.assert('C-a c - triggers, chat.input scope', count === 3);

         keymage.popScope('chat');
         fire({code: 65, ctrl: true}); fire({code: 65});
         wru.assert('C-a a - triggers, no scope', count === 4);
         fire({code: 65, ctrl: true}); fire({code: 66});
         wru.assert('C-a b - does not trigger, no scope', count === 4);

         keymage.pushScope('loading');
         keymage.popScope();
         wru.assert('popScope works', keymage.getScope() === '');
     }},

    {name: 'unbind',
     test: function() {
         var count = 0;
         var handler = function() { count++; };
         keymage('ctrl-a a', handler);
         keymage('chat.input', 'ctrl-a c', handler);

         keymage.pushScope('chat.input');
         fire({code: 65, ctrl: true}); fire({code: 65});
         fire({code: 65, ctrl: true}); fire({code: 67});
         wru.assert('both handlers triggered', count === 2);

         keymage.unbind('ctrl-a a', handler);
         fire({code: 65, ctrl: true}); fire({code: 65});
         wru.assert('nothing triggered', count === 2);

         keymage.unbind('chat.input', 'ctrl-a c', handler);
         fire({code: 65, ctrl: true}); fire({code: 67});
         wru.assert('nothing triggered', count === 2);
     }},

    {name: 'Prepared unbinder',
     test: function() {
         var count = 0;
         var unbinder = keymage('ctrl-a a', function () { count++; });
         fire({code: 65, ctrl: true}); fire({code: 65});

         wru.assert('handler triggered', count === 1);

         unbinder();
         fire({code: 65, ctrl: true}); fire({code: 65});
         wru.assert('nothing triggered', count === 1);
     }}
]);


test/phantom.js
var system = require('system');
var page = new WebPage;

page.onConsoleMessage = function(msg) {
    // cursor removed from phantomjs
    if (!/^\s+(?:\\|\/|\||\-)/.test(msg)) {
        console.log(msg.replace('\n', ''));
    }
};

page.open('about:blank', function(status) {
    if (status !== 'success') {
        return phantom.exit();
    }

    page.evaluate(function() {
        window.phantomExit = false;
        window.quit = function() {
            window.phantomExit = true;
        };
        window.require = function() {
            return {wru: window.wru};
        };
        window.global = window;
    });

    page.injectJs('keymage.js');
    page.injectJs('test/vendor/wru.console.max.js');
    page.injectJs('test/test.js');

    setInterval(function() {
        page.evaluate(function() {
            return window.phantomExit;
        }) && phantom.exit();
    }, 50);
});


keymage.js
/** @license
 * keymage.js - Javascript keyboard bindings handling
 * http://github.com/piranha/keymage
 *
 * (c) 2012-2016 Alexander Solovyov under terms of ISC License
 */

(function(define, undefined) {
define(function() {
    'use strict';

    var VERSION = '1.1.3';
    var isOsx = ~navigator.userAgent.indexOf('Mac OS X');

    // Defining all keys
    var MODPROPS = ['shiftKey', 'ctrlKey', 'altKey', 'metaKey'];
    var MODS = {
        'shift': 'shift',
        'ctrl': 'ctrl', 'control': 'ctrl',
        'alt': 'alt', 'option': 'alt',
        'win': 'meta', 'cmd': 'meta', 'super': 'meta',
                          'meta': 'meta',
        // default modifier for os x is cmd and for others is ctrl
        'defmod':  isOsx ? 'meta' : 'ctrl'
        };
    var MODORDER = ['shift', 'ctrl', 'alt', 'meta'];
    var MODNUMS = [16, 17, 18, 91];

    var KEYS = {
        'backspace': 8,
        'tab': 9,
        'enter': 13, 'return': 13,
        'pause': 19,
        'caps': 20, 'capslock': 20,
        'escape': 27, 'esc': 27,
        'space': 32,
        'pgup': 33, 'pageup': 33,
        'pgdown': 34, 'pagedown': 34,
        'end': 35,
        'home': 36,
        'ins': 45, 'insert': 45,
        'del': 46, 'delete': 46,

        'left': 37,
        'up': 38,
        'right': 39,
        'down': 40,

        '*': 106,
        '+': 107, 'plus': 107,
        'minus': 109,
        ';': 186,
        '=': 187,
        ',': 188,
        '-': 189,
        '.': 190,
        '/': 191,
        '`': 192,
        '[': 219,
        '\\': 220,
        ']': 221,
        "'": 222
    };

    var i;
    // numpad
    for (i = 0; i < 10; i++) {
        KEYS['num-' + i] = i + 95;
    }
    // top row 0-9
    for (i = 0; i < 10; i++) {
        KEYS['' + i] = i + 48;
    }
    // f1-f24
    for (i = 1; i < 25; i++) {
        KEYS['f' + i] = i + 111;
    }
    // alphabet
    for (i = 65; i < 91; i++) {
        KEYS[String.fromCharCode(i).toLowerCase()] = i;
    }

    // Reverse key codes
    var KEYREV = {};
    Object.keys(KEYS).forEach(function(k) {
        var val = KEYS[k];
        if (!KEYREV[val] || KEYREV[val].length < k.length) {
            KEYREV[val] = k;
        }
    });

    // -----------------------
    // Actual work is done here

    var currentScope = '';
    var allChains = {};

    function parseKeyString(keystring) {
        var bits = keystring.split(/-(?!$)/);
        var button = bits[bits.length - 1];
        var key = {code: KEYS[button]};

        if (!key.code) {
            throw 'Unknown key "' + button + '" in keystring "' +
                keystring + '"';
        }

        var mod;
        for (var i = 0; i < bits.length - 1; i++) {
            button = bits[i];
            mod = MODS[button];
            if (!mod) {
                    throw 'Unknown modifier "' + button + '" in keystring "' +
                        keystring + '"';
            }
            key[mod] = true;
        }

        return key;
    }

    function stringifyKey(key) {
        var s = '';
        for (var i = 0; i < MODORDER.length; i++) {
            if (key[MODORDER[i]]) {
                s += MODORDER[i] + '-';
            }
        }
        s += KEYREV[key.code];
        return s;
    }

    function normalizeKeyChain(keychainString) {
        var keychain = [];
        var keys = keychainString.split(' ');

        for (var i = 0; i < keys.length; i++) {
            var key = parseKeyString(keys[i]);
            key = stringifyKey(key);
            keychain.push(key);
        }

        keychain.original = keychainString;
        return keychain;
    }

    function eventKeyString(e) {
        var key = {code: e.keyCode};
        for (var i = 0; i < MODPROPS.length; i++) {
            var mod = MODPROPS[i];
            if (e[mod]) {
                key[mod.slice(0, mod.length - 3)] = true;
            }
        }
        return stringifyKey(key);
    }

    function getNestedChains(chains, scope) {
        for (var i = 0; i < scope.length; i++) {
            var bit = scope[i];

            if (bit) {
                chains = chains[bit];
            }

            if (!chains) {
                break;
            }
        }
        return chains;
    }

    var sequence = [];
    function dispatch(e) {
        // Skip all modifiers
        if (~MODNUMS.indexOf(e.keyCode)) {
            return;
        }

        var seq = sequence.slice();
        seq.push(eventKeyString(e));
        var scope = currentScope.split('.');
        var matched, chains, key;

        for (var i = scope.length; i >= 0; i--) {
            chains = getNestedChains(allChains, scope.slice(0, i));
            if (!chains) {
                continue;
            }
            matched = true;
            for (var j = 0; j < seq.length; j++) {
                key = seq[j];
                if (!chains[key]) {
                    matched = false;
                    break;
                }
                chains = chains[key];
            }

            if (matched) {
                break;
            }
        }

        var definitionScope = scope.slice(0, i).join('.');
        var preventDefault = chains.preventDefault;

        // partial match, save the sequence
        if (matched && !chains.handlers) {
            sequence = seq;
            if (preventDefault) {
                e.preventDefault();
            }
            return;
        }

        if (matched) {
            for (i = 0; i < chains.handlers.length; i++) {
                var handler = chains.handlers[i];
                var options = handler._keymage;

                var res = handler.call(options.context, e, {
                    shortcut: options.original,
                    scope: currentScope,
                    definitionScope: definitionScope
                });

                if (res === false || preventDefault) {
                    e.preventDefault();
                }
            }
        }

        // either matched or not, drop the sequence
        sequence = [];
    }

    function getHandlers(scope, keychain, fn) {
        var bits = scope.split('.');
        var chains = allChains;
        bits = bits.concat(keychain);

        for (var i = 0, l = bits.length; i < l; i++) {
            var bit = bits[i];
            if (!bit) continue;

            chains = chains[bit] || (chains[bit] = {});
            if (fn && fn._keymage.preventDefault) {
                chains.preventDefault = true;
            }

            if (i === l - 1) {
                var handlers = chains.handlers || (chains.handlers = []);
                return handlers;
            }
        }
    }

    function assignKey(scope, keychain, fn) {
        var handlers = getHandlers(scope, keychain, fn);
        handlers.push(fn);
    }

    function unassignKey(scope, keychain, fn) {
        var handlers = getHandlers(scope, keychain);
        var idx = handlers.indexOf(fn);
        if (~idx) {
            handlers.splice(idx, 1);
        }
    }

    function parsed(scope, keychain, fn, options) {
        if (keychain === undefined && fn === undefined) {
            return function(keychain, fn) {
                return keymage(scope, keychain, fn);
            };
        }

        if (typeof keychain === 'function') {
            options = fn;
            fn = keychain;
            keychain = scope;
            scope = '';
        }

        var normalized = normalizeKeyChain(keychain);

        return [scope, normalized, fn, options];
    }

    // optional arguments: scope, options.
    function keymage(scope, keychain, fn, options) {
        var args = parsed(scope, keychain, fn, options);
        fn = args[2];
        options = args[3];
        fn._keymage = options || {};
        fn._keymage.original = keychain;
        assignKey.apply(null, args);

        return function () {
            unassignKey.apply(null, args);
        };
    }

    keymage.unbind = function(scope, keychain, fn) {
        var args = parsed(scope, keychain, fn);
        unassignKey.apply(null, args);
    };

    keymage.parse = parseKeyString;
    keymage.stringify = stringifyKey;

    keymage.bindings = allChains;

    keymage.setScope = function(scope) {
        currentScope = scope ? scope : '';
    };

    keymage.getScope = function() { return currentScope; };

    keymage.pushScope = function(scope) {
        currentScope = (currentScope ? currentScope + '.' : '') + scope;
        return currentScope;
    };

    keymage.popScope = function(scope) {
        var i;

        if (!scope) {
            i = currentScope.lastIndexOf('.');
            scope = currentScope.slice(i + 1);
            currentScope = i == -1 ? '' : currentScope.slice(0, i);
            return scope;
        }

        currentScope = currentScope.replace(
            new RegExp('(^|\\.)' + scope + '(\\.|$).*'), '');
        return scope;
    };

    keymage.version = VERSION;

    window.addEventListener('keydown', dispatch);

    return keymage;
});
})(typeof define !== 'undefined' ? define : function(factory) {
    if (typeof module !== 'undefined') {
        module.exports = factory();
    } else {
        window.keymage = factory();
    }
});
